federation/pkg/controller/service/servicecontroller.go

```
@@ -0,0 +1,956 @@
+	// One service in one cluster with same hash
+	if err == nil && svc != nil && len(svc.Items) == 0 {
+		return svc.Items[0]
+	} else {
+		return nil
+	}
+}
+
+func (s *ServiceController) ensureClusterServices(service *v1.Service, client clientset.Clientset, clusterName string) ([]string, error) {
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		svc := getServiceByLabel(client, service)
+		if svc != nil {
+			// service exists
+			glog.V(4).Infof("service exists in cluster %v, trying to update", clusterName)
+			_, err = client.Core().Services(service.Namespace).UpdateStatus(service)
+			if err == nil {
+				// Update successfully, get LB IPs
+				return retrieveExternalIPs(client, service)
+			}
+		}
+
+		if errors.IsNotFound(err) {
+			// Create service if is not found
+			glog.Infof("Service '%s/%s' does not exist in %s, trying to create new",
+				service.Namespace, service.Name, clusterName)
+			_, err = client.Core().Services(service.Namespace).Create(service)
+			if err == nil {
+				return retrieveExternalIPs(client, service)
+			}
+		}
+
+		if errors.IsConflict(err) {
+			glog.V(4).Infof("Not persisting update to service '%s/%s' that has been changed since we received it: %v",
+				service.Namespace, service.Name, err)
+			// ignore and get LB info
+			return retrieveExternalIPs(client, service)
+		}
+
+		glog.Warningf("Failed to get LoadBalancerStatus for service '%s/%s' after creating it in cluster %s: %v",
+			service.Namespace, service.Name, clusterName, err )
+		time.Sleep(clientRetryInterval)
+	}
+	return err
+}
+
+func (s *ServiceController) ensureDNSRecords(clusterName string, service *v1.Service) error {
+	for _, zone := range s.zones {
+		ResourceRecordSets, ok := zone.ResourceRecordSets()
+		if !ok {
+			//zone name or cluster name?
+			rrsList, err := ResourceRecordSets.List()
+			if err != nil {
+				return fmt.Errorf("fail to get rrs")
```

Please don't suppress errors like this.  Return the original error (err), or as a minimum include the original error in your error message.